improver/precipitation_type/hail_fraction.py

```diff
@@ -0,0 +1,191 @@
+
+        Returns:
+            Hail fraction array.
+        """
+
+        # Hail size of 2 mm that always corresponds to a non-zero hail fraction.
+        hail_size_limit = 0.002
+        # Cloud condensation level temperature in Kelvin. If below this temperature,
+        # a hail fraction of 0 is set.
+        ccl_limit = 268.15
+        # Convective cloud top temperature in Kelvin. If above this temperature,
+        # a hail fraction of 0 is set.
+        cct_limit = 258.15
+
+        hail_fraction = np.interp(vertical_updraught.data, [5, 50], [0, 0.25]).astype(
+            np.float32
+        )
+        hail_fraction[
+            (hail_size.data > hail_size_limit) & (hail_fraction < 0.05)
+        ] = 0.05
+        hail_fraction[
+            (cloud_condensation_level.data < ccl_limit)
+            | (convective_cloud_top.data > cct_limit)
+            | (hail_melting_level.data > orography.data)
+        ] = 0
+        return hail_fraction
+
+    def _make_hail_fraction_cube(
+        self,
+        hail_fraction: np.ndarray,
+        vertical_updraught: Cube,
+        hail_size,
+        cloud_condensation_level,
+        convective_cloud_top,
+        hail_melting_level,
+    ) -> Cube:
+        """Create a cube containing the hail fraction array using information
+        from the other available cubes.
+
+        Returns:
+            Hail fraction cube.
+        """
+        hail_fraction_cube = create_new_diagnostic_cube(
+            "hail_fraction",
+            "1",
+            template_cube=vertical_updraught,
+            mandatory_attributes=generate_mandatory_attributes(
+                [
+                    vertical_updraught,
+                    hail_size,
+                    cloud_condensation_level,
+                    convective_cloud_top,
+                    hail_melting_level,
+                ],
+                model_id_attr=self.model_id_attr,
+            ),
+            data=hail_fraction,
+        )
+        return hail_fraction_cube
+
+    def process(
+        self,
+        vertical_updraught: Cube,
+        hail_size: Cube,
+        cloud_condensation_level: Cube,
+        convective_cloud_top: Cube,
+        hail_melting_level: Cube,
+        orography: Cube,
+    ) -> Cube:
+        """Calculates the hail fraction using the maximum vertical updraught,
+        the hail_size, the cloud condensation level temperature, the convective cloud
+        top temperature, the altitude of the hail to rain phase change and the
+        orography.
+
+        Args:
+            vertical_updraught: Maximum vertical updraught in m/s.
+            hail_size: Hail size in m.
+            cloud_condensation_level: Cloud condensation level temperature in K.
+            convective_cloud_top: Convective cloud top in K.
+            hail_melting_level: Altitude of the melting of hail to rain in metres.
+            orography: Altitude of the orography in metres.
```

Comment:
I don't like the reliance on the inputs being in SI units. I'd like to centralise these checks, but until I have time to do so, can you add simple `convert_units` calls?

Reply:
I've added `convert_units` calls as suggested.